File: Week6_Python/readability/readability.py
import re

# Get the text.
text = input("Text: ")

# Count the number of letters in the text.
num_letters = sum(x.isalpha() for x in text)

# Count the number of words in a sentence.
num_words = sum(x.isspace() for x in text) + 1
# Calculate the average letters per words.
avg_letters_count_per_words = num_letters / num_words
# Calculate the average number of letters per 100 words.
avg_letters_count_per_100_words = avg_letters_count_per_words * 100

# Count the number of sentences.
num_sentences = len(re.findall('[.!?]+', text))
# Calculate the average number of sentences per words.
avg_sentences_count_per_words = num_sentences / num_words
# Calculate the average number of sentences per 100 words.
avg_sentences_count_per_100_words = avg_sentences_count_per_words * 100

coleman_index = 0.0588 * avg_letters_count_per_100_words - 0.296 * avg_sentences_count_per_100_words - 15.8
index = round(coleman_index)

if index < 1:
    grade = "Before Grade 1"
elif index >= 16:
    grade = "Grade 16+"
else:
    grade = "Grade {0}".format(index)

print(num_letters, "letter(s)")
print(num_words, "word(s)")
print(num_sentences, "sentence(s)")
print("Average letters count per words:", avg_letters_count_per_words)
print("Average letters count per 100 words:", avg_letters_count_per_100_words)
print("Average sentences count per words:", avg_sentences_count_per_words)
print("Average sentences count per 100 words:", avg_sentences_count_per_100_words)
print(grade)

File: Week2_Arrays/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int main(int argc, string argv[])
{
    string text = get_string("Text: ");
    int letters = 0;
    int words = 0;
    int sentences = 0;

    for (int i = 0, n = strlen(text); i < n; i++)
    {
        char letter = text[i];

        // Count the number of letters.
        if (isalpha(letter))
        {
            letters = letters + 1;
        }

        // Count the number of words.
        if (isspace(letter))
        {
            words = words + 1;
        }

        int is_end_of_sentence = text[i] == '.' || text[i] == '!' || text[i] == '?';

        // Count the number of sentences.
        if (is_end_of_sentence)
        {
            sentences = sentences + 1;
        }
    }

    words = words + 1;

    // Calculate the Coleman-Liau index.
    float calculatedIndex = (0.0588 * letters / words * 100) - (0.296 * sentences / words * 100) - 15.8;
    int index = round(calculatedIndex);

    // Output the correct statement based on the above calculated index.
    if (index < 0)
    {
        printf("%s\n", "Before Grade 1");
    }
    else if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("%s %i\n", "Grade", index);
    }
}